### ISSUE TITLE
How can I train only the classes I want?

### ISSUE THREAD

**Issue by nameCDI** [ultralytics/ultralytics#1541]
### Search before asking

- [X] I have searched the YOLOv8 [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/ultralytics/ultralytics/discussions) and found no similar questions.

### Question

For example, if my dataset has classes 0-9, can I select only classes [0,1,3,7,8] to train? 

Is it possible to do this without modifying the dataset and editing the code? 

I think modifying the dataloader might work, but I’m not sure which part to modify.

Thank you.

**Comment by Laughing-q** [comment]
@nameCDI just add `classes=[0,1,3,7,8]`
```python
model.train(....., classes=[0,1,3,7,8])
```

**Comment by nameCDI** [comment]
@Laughing-q 

Thank you a lot for your help! It was amazing to me.

But I have another question.

<img width="944" alt="train_nc=11" src="https://user-images.githubusercontent.com/128284613/226863796-51115f97-986c-42af-ad16-b25ee3b7e6d7.png">

In my task, Overriding model.yaml nc=80 with nc=11.

It's different from my number of classes nc=6.

<img width="442" alt="docs_nc" src="https://user-images.githubusercontent.com/128284613/226864750-6680e4f7-e0b4-4ace-8e95-b66d8aecb879.png">

[https://docs.ultralytics.com/reference/ops/#ultralytics.yolo.utils.ops.non_max_suppression]

Should I fix this like 'nc=6'?

**Comment by Laughing-q** [comment]
@nameCDI I suggest you to filter the dataset first if you want the channel of `nc` in model architecture equals to 6. 
classes arg is lightweight and simply ignore classes that are not included in the classes list, during train, val and predict, it has no effect on model architecture.

**Comment by nameCDI** [comment]
> @nameCDI I suggest you to filter the dataset to include the 6 classes you want to train if you want the channel of `nc` in model architecture equals to 6. classes arg is lightweight and simply ignore classes that are not included in the classes list, during train, val and predict, it has no effect on model architecture.

Thanks a lot.  I understood.
Then it would be better to filter dataset if I want to fine tune after training.

**Comment by Laughing-q** [comment]
@nameCDI closing the issue now! please feel free to reopen it if you have related issues. :)

**Comment by aloideao** [comment]
@Laughing-q  what about terminal training? I tried doing that just like yolov5 in the dataloader but it didnt work 
https://github.com/ultralytics/ultralytics/blob/ff91fbd9c391bf7e46593ed06df4436e6555dcde/ultralytics/yolo/data/dataloaders/v5loader.py#L549

**Comment by glenn-jocher** [comment]
@aloideao thank you for reaching out. In YOLOv8, when training in the terminal, you can pass your classes list to `--hyp` parameter with a `data/hyp.scratch.yaml` file (you may also use `--classes` parameter to specify classes). In this yaml file, add `classes` field with your classes listed, like the following:

```yaml
# path: data/hyps/hyp.scratch.yaml
...
classes: [0, 1, 3, 7, 8]
```

However, nonstandard paths for `--data` or `--cfg` files currently require modification of Ultralytics scripts, such as the 'parser_add_yolo_layers.py' function in 'utils/README.md'. You can also set command-line arguments, like `--cfg data/my_config.yml --weights data/weights/last.pt --img 640 ...`, directly. Please let me know if there's anything else you need help with.

**Comment by Laughing-q** [comment]
@aloideao just pass `classes` flag, like if you only want to train your model with the first two classes, then add `classes=[0,1]`.

**Comment by karthikyerram** [comment]
<img width="906" alt="image" src="https://github.com/ultralytics/ultralytics/assets/144154810/b00631bb-8c81-4fc2-b468-bf4ef3fbb163">
When I do that, I get this error

**Comment by glenn-jocher** [comment]
@karthikyerram i apologize for any confusion. There seems to be a misunderstanding; on the terminal (command line), you cannot directly filter classes using a simple flag in YOLOv8, as you would in Python code. To train on specific classes, you typically need to filter your dataset first, so that it only contains the classes you want to train on.

To clarify, you would edit your dataset's annotation files to only include the desired classes `[0,1,3,7,8]`. This means removing any annotations for classes you do not want to include and possibly renumbering your classes so they are sequential. For instance, if you remove class '2', the class '3' should be renamed to '2' in your annotations and label files, and so on, to maintain a continuous sequence of class IDs.

After you've filtered and re-labeled your dataset appropriately, you should also adjust your `model.yaml` to have `nc` (number of classes) to match the actual number of classes you're training on, in your case, that would be 5.

This preparatory work of dataset curation is necessary because the YOLOv8 model configuration and training pipeline expect consistent and properly labeled data to train effectively.

**Comment by aitor-martinez-seras** [comment]
@glenn-jocher Read the issue and other related issues (#1558) . Just here to note that it would be amazing if YOLOv8 could include this feature of filtering out the classes you want to train on directly with an argument on the config (for example).  Replicating the dataset and its labels in other folder (at least if you want train on less classes and test in all clases to test, for example, open world scenarios) with only the desired annotations its a time-consuming and memory intensive task that could be alleviated with that implementation. 

Are you planning on including it? 

In any case, thank you for work! Keep it up!

**Comment by expen6ve** [comment]
> [@karthikyerram](https://github.com/karthikyerram) i apologize for any confusion. There seems to be a misunderstanding; on the terminal (command line), you cannot directly filter classes using a simple flag in YOLOv8, as you would in Python code. To train on specific classes, you typically need to filter your dataset first, so that it only contains the classes you want to train on.
> 
> To clarify, you would edit your dataset's annotation files to only include the desired classes `[0,1,3,7,8]`. This means removing any annotations for classes you do not want to include and possibly renumbering your classes so they are sequential. For instance, if you remove class '2', the class '3' should be renamed to '2' in your annotations and label files, and so on, to maintain a continuous sequence of class IDs.
> 
> After you've filtered and re-labeled your dataset appropriately, you should also adjust your `model.yaml` to have `nc` (number of classes) to match the actual number of classes you're training on, in your case, that would be 5.
> 
> This preparatory work of dataset curation is necessary because the YOLOv8 model configuration and training pipeline expect consistent and properly labeled data to train effectively.

What if i want to do a transfer learning on a pretrained yolov8 into a custom motorcycle dataset. how do i make it so that it would only be trained or be used for motorcycles which is "3" in the classes?

**Comment by glenn-jocher** [comment]
@expen6ve to fine-tune a pretrained YOLOv8 model on a custom dataset for "motorcycles" (class 3), you can set your dataset annotations to include only the required "motorcycle" class. Ensure the dataset contains only the annotations for motorcycles, relabel the class as `0` if it's the sole class in the dataset, and update the dataset YAML file to reflect this change (`nc=1` for one class). 

During training, you can use the `--data` argument to specify your custom dataset YAML and the `--weights` argument to initialize with a pretrained YOLOv8 model, e.g.:

```bash
yolo task=detect mode=train model=yolov8n.pt data=custom_motorcycle.yaml epochs=50
``` 

Feel free to refer to [Ultralytics training guide](https://docs.ultralytics.com/modes/train/) for more details.
